Authentication for developers (#28)
City-of-Helsinki/bestpractice: docs/authentication.md

> # Authentication

Reviewer: ```suggestion
---
name: Authentication
menu: Best practices
route: /authentication
---

# Authentication
```

Most of the projects requiring any sort of user authentication should use Tunnistamo.

> Tunnistamo is service that authenticates users against several different identity sources,

Reviewer: ```suggestion
Tunnistamo is a service that authenticates users against several different identity sources,
```

including Espoo, Vantaa and Helsinki internal accounts, Facebook, Google and Github.
Tunnistamo also keeps a running session for the user, meaning the user won't be needing
to login again to access another service using tunnistamo.

The actual line protocols are OAuth2 and OIDC. New applications should always use OIDC.

*Nomenclature alert*: Tunnistamo will soon be changing names, as the implementation will switch from your own implementation (named "tunnistamo") to Keycloak based implementation. Keeping the same name would be too confusing. We're not sure of the upcoming name yet.

## Getting necessary Tunnistamo identifiers

> To get started you will either need a locally setup instance of Tunnistamo or get access to one of shared instances available within Helsinki infrastructure. I'd recommend that you try and get access to the shared instances, unless you must develop offline. 

Reviewer: ```suggestion
To get started you will either need a locally setup instance of Tunnistamo or get access to one of shared instances available within Helsinki infrastructure. It is recommended that you try and get access to the shared instances, unless you must develop offline. 
```

Current instances are:
* Kanslia (executive office) instance: default instance used by projects, both during development and production
* KuVa (culture and leisure) instance: instance used by KuVa projects during their development. Very flexible registration

> To get access to Kanslia instance, pop by the slack channel #tunnistamo

Reviewer: ```suggestion
To get access to Kanslia instance, pop by the slack channel #tunnistamo.
```
